### PR TITLE
[GeoMechanicsApplication] Refactor processing of DOFs by elements and conditions

### LINKED ISSUE #12073: [GeoMechanicsApplication] Refactor processing of DOFs by elements and conditions
As a developer, I would like to define the order of degrees of freedom (in elements and conditions) just once, instead of repeating it over and over again. This will improve the maintenance of the corresponding code. Furthermore, it will become much easier to change the order if needed. In fact, when this issue has been addressed, we would like to change the DOF order of U-Pw same order elements to match the order adopted by U-Pw diff order elements. That is, however, the scope of [another issue](https://github.com/KratosMultiphysics/Kratos/issues/12061).

### PR DESCRIPTION
**Pull request by avdg81**
**📝 Description**
This PR eliminates several repetitions of the order of degrees of freedom (in elements and conditions). Ideally, the order should be defined just once. This will improve the maintenance of the corresponding code. Also, it will become much easier to change the order as needed.

As a bonus, due to removing the duplications, we have also addressed a significant amount of code smells.

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: To avoid conflicts with other applications, I'd propose to make this a separate namespace
```suggestion
namespace Kratos::DofUtilities
```

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: I think this could be const (or does the transform take a non-const ref?)
```suggestion
    const auto get_dof_value = [Step](const auto p_dof) -> double {
```

File: applications/GeoMechanicsApplication/custom_conditions/T_condition.h
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_conditions/T_condition.h]: In Pw_condition the GetDofs implementation is in the .cpp file ( for me the expected place. ) Here it is in the .h file.
Their implementation is necessarily very similar. I wonder whether it can be a utility implementation where the D.O.F. variable name or a list of variable names is passed ( WATER_PRESSURE, TEMPERATURE etc. )

File: applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp]: Step here is probably 1 for start of the step ( previous value ), 0 for end of the step ( current value ). Apart from the confusing inverse setup of 0 and 1, maybe the name Step can be improved.

File: applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp:76]: I wonder whether this should have an implementation or should it be abstract and always overridden by a specific implementation.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp:653]: This will probably be the only remaining implementation for the UPw element when the whole redesign is complete.

File: applications/GeoMechanicsApplication/custom_elements/geo_structural_base_element.cpp
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_elements/geo_structural_base_element.cpp:482]: Which makes clear that the structural base element is a beam element.

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: On first glance these are four repetitions of a similar thing. The remark about VariableData I don't understand yet. As far as I have understood all D.O.F. are double by nature.

**Review by WPK4FEM — COMMENTED** [review]
Hi Anne,
That condensed functionality considerably, thank you. Some parts left that I don't properly understand, maybe spoken explanation helps there.
Wijtze Pieter

File: applications/GeoMechanicsApplication/custom_conditions/Pw_condition.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_conditions/Pw_condition.cpp]: It's nice to see these examples of the stl library

File: applications/GeoMechanicsApplication/custom_conditions/T_condition.h
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_conditions/T_condition.h]: Could we move this function to the cpp?

File: applications/GeoMechanicsApplication/custom_conditions/T_condition.h
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_conditions/T_condition.h]: Could we move this function to the cpp?

File: applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp]: Seeing the inheritance structure (and since `GetValuesVector` is an override from `Condition`), I think it would be better if we move this function to the U_Pw Condition (which is way easier due to your refactoring). However, it is of course a bit out of scope, so I leave it up to you if you think it's good to do it in this PR.

File: applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp]: I think this loop for the water pressure could be another transform with back_inserter

File: applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp:96]: I think we should not keep these comments (but do we have an answer?)

File: applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp]: As @WPK4FEM said, there are indeed quite some versions of getting a 'single' dof (like WATER_PRESSURE/TEMPERATURE), we might extract a function.

File: applications/GeoMechanicsApplication/custom_elements/transient_Pw_interface_element.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/transient_Pw_interface_element.cpp:709]: Also here, it might be better to remove the comments

File: applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h]: Here it can also be an std::transform (or an extracted function)

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp:37]: since `EquationId()` is a const function, I think also the Dof<double>* could be const here. Same holds for the other functions if I'm correct.

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: Also here I'm not sure if we should keep the comment

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: If we extract a function, we could extract the first time derivative `Variable` using the cast, and for the second just call `GetTimeDerivative` on the extracted first time derivative. This saves some duplication, especially since we have to do some extra explanation in the comment I think it would help to have that only in one place. We can still create the lambdas for the transforms

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_dof_utilities.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_dof_utilities.cpp:108]: We are still missing tests for the other 6 utility functions

**Review by rfaasse — CHANGES_REQUESTED** [review]
Very impressive @avdg81, this cleans up so many duplications (see also the net negative line count 😄 ) and gives nice examples of how to use the stl library: 
image

I have some suggestions for improvements and minor clean-up. 

In my opinion. the only one thing that's still missing before we can merge are unit tests for some of the new utility functions.

File: applications/GeoMechanicsApplication/custom_conditions/T_condition.h
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_conditions/T_condition.h]: Yes, I have created a utility function that returns the degrees of freedom for the given collection of nodes and a single DOF variable. I have not implemented the same functionality given a _list_ of DOF variables. The reason is that in the current situation we need to accommodate two orderings for U-Pw elements: (1) where displacements and water pressures are intertwined [for the non-diff order elements], and (2) where all displacement DOFs come first followed by the water pressure DOFs [for the diff order elements]. I'd propose to revisit this after we have made the DOF ordering consistent across all U-Pw elements.

Also, I did move the implementation from the header file to the implementation file, as you suggested.

File: applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp]: I agree with you that a name such as `BufferIndex` would be much more accurate and descriptive. However, we inherit this parameter name from the Kratos core (see `condition.h`). So if we would like to change it, I feel we'd better start from there. I did, however, adopt the parameter name `BufferIndex` while writing the new DOF utility functions.

Also, there is an issue with the virtual function itself, since it has a default value for parameter `Step`, which opens the door to having different default values for different overrides. See for instance [this rule of the C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c140-do-not-provide-different-default-arguments-for-a-virtual-function-and-an-overrider), which warns for that. Clang-tidy also reports it:

image

File: applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp:76]: As indicated before, we might be able to simplify things even more after we have rearranged the DOFs. But since we intend to replace these element classes (as part of the element redesign), we'd better do this right when we implement the new elements.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp:653]: Yes, that makes sense.

File: applications/GeoMechanicsApplication/custom_elements/geo_structural_base_element.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/geo_structural_base_element.cpp:482]: Agreed. This is something to keep an eye on when we move over to the structural elements of the `StructuralMechanicsApplication` (since the "geo" structural elements will be removed at some point in time).

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: We have discussed this in-person. The bottom line is that the functions ending with `OfUPwDofs` are only there to not change the current behavior. However, we do need to investigate why the water pressure (and its time derivatives) are assumed to be zero. When we can show that this is not needed at all (or if we can find a better alternative), then we can start changing the function's behavior. And, in fact, that may mean we no longer need these dedicated functions.

Also, I have rephrased the comment about the `VariableData`. Could you please check whether it is clearer now?

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: Good suggestion. I have made the namespace even more specific: `Kratos::Geo::DofUtilities`.

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: The `auto` here deduces the type of the lambda expression. So far, I have never encountered a lambda expression that was assigned to a `const auto` instance. I have browsed through the examples in the C++ Core Guidelines, but they all seem to use just `auto`, so without the `const`. I would suggest to follow those examples, unless you can provide me with an authoritative reference that actually prescribes to use `const auto` for these cases.

File: applications/GeoMechanicsApplication/custom_conditions/Pw_condition.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_conditions/Pw_condition.cpp]: Thank you :-)

File: applications/GeoMechanicsApplication/custom_conditions/T_condition.h
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_conditions/T_condition.h]: Wijtze Pieter had the exact same remark, and I totally agree with it. Done.

File: applications/GeoMechanicsApplication/custom_conditions/T_condition.h
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_conditions/T_condition.h]: Good catch. Yes, we can. Done.

File: applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp]: Good suggestion. I would, however, do that in a separate PR, since it might change existing behavior. The aim of this PR is a pure refactoring, so no behavioral changes.

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: I was indeed to quick with this comment, I had the same comment for the lambdas in the dof utils, but came to the same conclusion that it does not make too much sense for a lambda expression and removed them again. Totally agree with you that we keep it as is

File: applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp]: Done. In fact, I have moved that code (with `std::transform`) to a utility function and call it from here.

File: applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp:96]: The problem is that at present we don't know why the water pressures (and their time derivatives) are set to zero by these functions. In my opinion, that is beyond the scope of this PR. We should, however, create an issue to investigate it. These new comments serve as a reminder and as a way to quickly find these anomalies when we attempt to address them.

File: applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp]: Thank you for the suggestion. I have added a utility function (`ExtractDofsFromNodes`) that returns the degrees of freedom of a collection of nodes given a DOF variable.

File: applications/GeoMechanicsApplication/custom_elements/transient_Pw_interface_element.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/transient_Pw_interface_element.cpp:709]: See my earlier explanation of why I would like to keep them.

File: applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h]: Yes, I have reused the extracted function (`ExtractDofsFromNodes`) here. I have also searched the other files that I have touched in this PR, and called this new function from there as well.

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp:37]: No, unfortunately, the alias `DofsVectorType` (see, for instance, `element.h`) uses raw pointers to mutable objects:
```
typedef std::vector<DofType::Pointer> DofsVectorType;
```
where `typedef Dof<double> DofType;` and `using Pointer=Dof*;`. So I cannot change this to `std::vector<const Dof<double>*>`. However, as you can see in the parameters of the lambda expressions, I do assign each pointer-to-mutable to a pointer-to-const: `[Step](const auto p_dof) -> double { ... }`. I don't see how I can improve this further. If you have any ideas, please share them with me. Thanks.

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: See my previous replies.

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: Good suggestion. Done.

**Review by rfaasse — DISMISSED** [review]

**Review by rfaasse — COMMENTED** [review]
Trying to add a new review to remove blocking the PR due to 'Request Changes', but also not approving it.

File: applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp]: Agreed!

File: applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp:96]: Good point, let's keep them then indeed!

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp:37]: Ahh this rings a bell, that is indeed unfortunate, Indeed since you use const in the lambdas it is the best we can do given the `DofsVectorType`.

File: applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h]: Thanks, looks very clean!

File: applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp]: Perfect

File: applications/GeoMechanicsApplication/custom_elements/transient_Pw_interface_element.cpp
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_elements/transient_Pw_interface_element.cpp:709]: Agreed

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: Nice!

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: For readability, it might be nice to add parentheses here (at first I was wondering why we need the adres of pDof, but it's the pointer to the variable).
```suggestion
    auto p_variable = dynamic_cast<const Variable<double>*>(&(pDof->GetVariable()));
```

**Review by rfaasse — COMMENTED** [review]
Thank you for processing the review comments! The quality improvement with respect to the original code (i.e. current master branch) is really impressive! 

I have a single small nitpick about parentheses we can add for readability, but it's really minor and apart from the unit tests this branch is ready to be merged from my point of view.

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp]: Yes, I agree that parentheses make the intention clearer. I have added them.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_dof_utilities.cpp
Comment (avdg81) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_dof_utilities.cpp:108]: I have added new unit tests to cover these new utility functions.

File: applications/GeoMechanicsApplication/custom_utilities/dof_utilities.h
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_utilities/dof_utilities.h:42]: Very nice generalization!

**Review by rfaasse — APPROVED** [review]
Thank you for your hard work on the tests, they are very easy to read and cover the extracted functionality nicely! From my point of view this PR is ready to be merged!

**Review by WPK4FEM — APPROVED** [review]
Have read through all. Understood most and am happy with removal of multiple definitions D.O.F. and equationId ordering.

**Pull request merged**
